### PR TITLE
Remove extra options to manipulate `JsonSerializerSettings`

### PR DESCRIPTION
**Pull request by dougbu**
- #4339: remove non-recommended JSON formatter constructors
 - affects `JsonInputFormatter`, `JsonOutputFormatter`, `JsonPatchInputFormatter`
 - `JsonOutputFormatter` cleanup also impacts `JsonHelper`
 - make `SerializerSettingsProvider` class public and use it
- #4409: make `SerializerSetings` properties get-only and `protected`
 - affects `JsonInputFormatter`, `JsonOutputFormatter`

Recommended patterns:
- change `JsonSerializerSettings` values in `MvcJsonOptions` for almost all customizations
- find `JsonOutputFormatter` in `MvcOptions.OutputFormatters` when limiting per-result formatters
- start with `SerializerSettingsProvider.CreateSerializerSettings()` when customizing a per-result formatter

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs [hunk 1]
```diff
@@ -25,16 +24,11 @@ public class JsonOutputFormatter : TextOutputFormatter
...
 
-        public JsonOutputFormatter()
-            : this(SerializerSettingsProvider.CreateSerializerSettings(), ArrayPool<char>.Shared)
-        {
-        }
-
-        public JsonOutputFormatter(JsonSerializerSettings serializerSettings)
-            : this(serializerSettings, ArrayPool<char>.Shared)
-        {
-        }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs]: Removing these wasn't mentioned in #4339 but seems to fit. Let me know if you disagree.

File: test/WebSites/ApiExplorerWebSite/Startup.cs [hunk 2]
```diff
@@ -23,9 +23,23 @@ public void ConfigureServices(IServiceCollection services)
...
+                JsonOutputFormatter jsonOutputFormatter = null;
+                for (var i = 0; i < options.OutputFormatters.Count; i++)
+                {
+                    var formatter = options.OutputFormatters[i];
+                    jsonOutputFormatter = formatter as JsonOutputFormatter;
+                    if (jsonOutputFormatter != null)
+                    {
+                        break;
+                    }
+                }
```
Comment (dougbu) [inline test/WebSites/ApiExplorerWebSite/Startup.cs]: Thoughts on adding a `FormatterCollection<TFormatter>.Find<T>() where T : TFormatter` method (or `Get()` or `FindType()` or ...)? I did this twice in this PR and that may indicate users would do the same.

File: test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs [hunk 3]
```diff
@@ -1,28 +1,70 @@
...
 namespace BasicWebSite.Controllers.ContentNegotiation
 {
     public class NormalController : Controller
     {
+        private JsonOutputFormatter _indentingFormatter;
+
+        public NormalController(IOptions<MvcJsonOptions> jsonOptions, ArrayPool<char> charPool)
+        {
+            var defaultSettings = jsonOptions.Value.SerializerSettings;
+            var settings = new JsonSerializerSettings
```
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs]: This is tedious but `JsonSerializerSettings` lacks a copy constructor. Any objection to this test site simply using `new JsonSerializerSettings { Formatting = Formatting.Indented }`? Middle ground would be to copy over only what `SerializerSettingsProvider` changes.

File: test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs [hunk 4]
```diff
@@ -31,7 +33,9 @@ private class ShortCircuitWithFormatterAttribute : Attribute, IResourceFilter
 
             public ShortCircuitWithFormatterAttribute()
             {
-                _formatters = new IOutputFormatter[] { new JsonOutputFormatter() };
+                _formatters = new IOutputFormatter[] { new JsonOutputFormatter(
+                    new JsonSerializerSettings(),
```
Comment (dougbu) [inline test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs]: Hmm, should I do the Right Thing<sup>:tm:</sup> and get the settings from `IOptoins<MvcJsonOptions>`? That is, is it :ok: to demonstrate the lazy approach in one of our web sites? Same question for `JsonFormatterController` below.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs (hunk 1 above)
Comment (kichalla) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs]: Users trying to do per-controller specific formatters was fairly common in Web API. In this scenario users would have to new up both input and output formatters. Removing this probably does not give the best experience IMO.

File: test/WebSites/FormatterWebSite/Controllers/JsonFormatterController.cs [hunk 5]
```diff
@@ -21,8 +22,11 @@ public IActionResult ReturnsIndentedJson()
                 Name = "John Williams"
             };
 
-            var jsonFormatter = new JsonOutputFormatter();
-            jsonFormatter.SerializerSettings.Formatting = Formatting.Indented;
+            var settings = new JsonSerializerSettings
+            {
+                Formatting = Formatting.Indented,
+            };
+            var jsonFormatter = new JsonOutputFormatter(settings, ArrayPool<char>.Shared);
```
Comment (kichalla) [inline test/WebSites/FormatterWebSite/Controllers/JsonFormatterController.cs]: This is a fairly common scenario. Users needing to know about `ArrayPool` is probably not a good experience.

File: test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs (hunk 4 above)
Comment (kichalla) [inline test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs]: (I suck at these emojis :-)...I did a +1 unintentionally...how do i revert that back?...doing  a -1 created one more entry :-))
The instance of serializer settings on `MvcJsonOptions` is shared across all controllers, so it cannot be used on `JsonFormatterController` as it is mutating it. Probably have a comment here about `MvcJsonOptions` here.

File: test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs (hunk 3 above)
Comment (kichalla) [inline test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs]: This is what I was not comfortable about with not having a way to create `JsonSerializerSettings` with all the defaults that the framework provides. I think we should provide an easier way for end users to get these settings.

File: test/WebSites/ApiExplorerWebSite/Startup.cs (hunk 2 above)
Comment (kichalla) [inline test/WebSites/ApiExplorerWebSite/Startup.cs]: Use LINQ instead?

Comment (dougbu) [inline test/WebSites/ApiExplorerWebSite/Startup.cs]: > Use LINQ instead?

Ick

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs]: > In this scenario users would have to new up both input and output formatters.

In what scenario would users need a per-controller _input_ formatter? And, despite our penchant for testing corner cases, when would users really need custom settings for a single output formatter?

File: test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs (hunk 3 above)
Comment (dougbu) [inline test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs]: > we should provide an easier way for end users to get these settings.

These settings come from very few places: `SerializerSettingsProvider.CreateSerializerSettings()` and the user's own code in `Startup` or a resource or action filter e.g. an `OnActionExecuting()` override in their controller. If needed, we could move `SerializerSettingsProvider` into a non-`.Internal` namespace.

File: test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs (hunk 4 above)
Comment (dougbu) [inline test/WebSites/FiltersWebSite/Controllers/ResourceFilterController.cs]: I wasn't suggesting reusing the global settings in `JsonFormatterController`, just grabbing them as a starting point. In any case, the user is not mutating the settings here in `ResourceFilterController`. In `JsonFormatterController`, would need to either leave the code as I've changed or match whatever copying we choose in `NormalController` (above).

Separately the previous code in `JsonFormatterController` is fragile. Should not depend on the formatter's constructor to create a unique `JsonSerializerSettings` instance. Another good reason for #4409, killing the `SerializerSettings` property to death.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs]: That said, if others agree, I could restore the `JsonOutputFormatter(JsonSerializerSettings)` overload. Don't see any reason for the parameterless constructor.

File: test/WebSites/FormatterWebSite/Controllers/JsonFormatterController.cs (hunk 5 above)
Comment (dougbu) [inline test/WebSites/FormatterWebSite/Controllers/JsonFormatterController.cs]: I doubt users often need to change serializer settings outside their `Startup` code. But, as I [said](https://github.com/aspnet/Mvc/pull/4492/files#r59992980) in `JsonOutputFormatter`, I could restore the `JsonOutputFormatter(JsonSerializerSettings)` constructor overload.

File: test/WebSites/ApiExplorerWebSite/Startup.cs (hunk 2 above)
Comment (javiercn) [inline test/WebSites/ApiExplorerWebSite/Startup.cs]: OfType<T>().FirstOrDefault(). I don't feel there's a need for a specifc method.

File: test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs (hunk 3 above)
Comment (javiercn) [inline test/WebSites/BasicWebSite/Controllers/ContentNegotiation/NormalController.cs]: I wouldn't go to all the trouble of copying all the settings if the intent of the test is to prove that you can override the formatter per controller, I would just set anything that is not a default value.

File: src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Formatters.Json/JsonOutputFormatter.cs]: :see_no_evil: Decided creating a customized JSON output formatter is an uncommon scenario and doesn't need to be incredibly simple. The required parameters are easy to create (because `SerializerSettingsProvider` is public) or to grab from DI. The same is true for the JSON input formatters though they have a couple more values that users must grab from DI.

**Comment by dougbu** [comment]
:up::date: also generalized the PR title and description

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/JsonHelper.cs [hunk 6]
```diff
@@ -46,7 +57,7 @@ public IHtmlContent Serialize(object value, JsonSerializerSettings serializerSet
                 throw new ArgumentNullException(nameof(serializerSettings));
             }
 
-            var jsonOutputFormatter = new JsonOutputFormatter(serializerSettings);
+            var jsonOutputFormatter = new JsonOutputFormatter(serializerSettings, _charPool);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/JsonHelper.cs]: :heart_eyes_cat:

**Comment by rynowak** [comment]
It would be good to document as part of the docs for these constructors have to get both the *app-wide* and *secure-default* serializer settings.

Other than that :shipit:

**Comment by dougbu** [comment]
:up::date: @rynowak I made the change we discussed offline...

**Comment by rynowak** [comment]
:shipit:

**Comment by dougbu** [comment]
d8d2e54
